[expo-print][ios] fix web urls and data strings support (again)

mroswald opened the pull request:
# Why

I may be wrong, but it seems that the recent [refactoring to Swift](https://github.com/expo/expo/pull/21561) introduced a regression to the `printAsync` function. I could not simply upgrade to `expo-print` 12.3.0 without errors, so I started investigating.

Error and way to reproduce it are described in the issue: https://github.com/expo/expo/issues/22971

Disclaimer: I'm not an expert with Swift and happy for any feedback on this! Nor I am very sure this is really an issue with the code or if I simply missed something else. I used ChatGPT for coding assistance.

# How

I extended the data fetching function to also work with http(s) urls and data strings (base64 encoded).

# Test Plan

Use this snack: https://snack.expo.dev/@maki123/print-usage
It's important to note, that the current Expo Go app will not reproduce the issue, as it's bundled with the "old" EXPrint framework prior to the change that introduced the regression.
Download the snack to a local project and run it. Make sure to use `expo-print` >= 12.3.0

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

mroswald committed:
Update ExpoPrintWithPrinter.swift

mroswald committed:
Update ExpoPrintWithPrinter.swift

fix indentation

mroswald committed:
Update ExpoPrintWithPrinter.swift

mroswald committed:
Update ExpoPrintWithPrinter.swift

behenate commented on packages/expo-print/ios/ExpoPrintWithPrinter.swift:
```diff
@@ -11,15 +11,16 @@ public class ExpoPrintWithPrinter {
-      guard let printingData = dataFromUri(uri: uri) else {
-        promise.reject(InvalidUrlException())
-        return
+      dataFromUri(uri: uri) { [self] data in
+        if let printingData = data {
+          self.printWithData(printingData: printingData, options: options, promise: promise)
```
We should use weak reference in here: 
```suggestion
      dataFromUri(uri: uri) { [weak self] data in
        if let self = self, let printingData = data {
          self.printWithData(printingData: printingData, options: options, promise: promise)
        ```

behenate commented on packages/expo-print/ios/ExpoPrintWithPrinter.swift:
```diff
@@ -137,11 +138,39 @@ public class ExpoPrintWithPrinter {
-      return try Data(contentsOf: URL(fileURLWithPath: uri))
-    } catch {
-      return nil
+  private func dataFromUri(uri: String, completion: @escaping (Data?) -> Void) {
+    if uri.hasPrefix("http") {
+      // Handle HTTP URL asynchronously
+      DispatchQueue.global().async {
+        if let url = URL(string: uri), let data = try? Data(contentsOf: url) {
+          DispatchQueue.main.async {
+            completion(data)
+          }
+        } else {
+          DispatchQueue.main.async {
+            completion(nil)
+          }
+        }
+      }
+    } else if uri.hasPrefix("data:") {
+      // Handle base64 encoded data URI
+      if let commaIndex = uri.firstIndex(of: ",") {
+        let base64String = String(uri[uri.index(after: commaIndex)...])
+        if let data = Data(base64Encoded: base64String) {
+          completion(data)
+        } else {
+          completion(nil)
+        }
+      }
+    } else {
+      // Handle local file path synchronously
+      let adjustedUri = uri.hasPrefix("file://") ? String(uri.dropFirst(7)) : uri
+
+      if let data = try? Data(contentsOf: URL(fileURLWithPath: adjustedUri)) {
+        completion(data)
+      } else {
+        completion(nil)
+      }
```
URL can handle the data and "file://" prefixes so we can reduce the number of ifs. Also we should use `URLSession` for http requests. In the variable declarations at the top of the file we should add:
```
  let printURLSession = URLSession(configuration: .default)
  var dataTask: URLSessionDataTask?
```
Here is an improved version of the function:
```suggestion
   private func dataFromUri(uri: String, completion: @escaping (Data?) -> Void) {
    guard var url = URL(string: uri) else {
      completion(nil)
      return
    }
    
    // Assume that URLs without a scheme eq. /home/user/file.pdf will be local file urls
    if url.scheme == nil {
      url = URL(fileURLWithPath: uri)
    }
    
    // process http and https requests asynchronously
    if url.scheme == "http" || url.scheme == "https" {
      dataTask = printURLSession.dataTask(with: URLRequest(url: url)) { [weak self] data, response, error in
        defer {
          self?.dataTask = nil
        }
        
        if error == nil, let response = response as? HTTPURLResponse, response.statusCode == 200 {
          DispatchQueue.main.async {
            completion(data)
          }
          return
        }
        completion(nil)
      };
      dataTask?.resume()
    } else {
      let data = try? Data(contentsOf: url)
      completion(data)
    }
  }
```

behenate reviewed (changes requested):
Thanks for the bug report and the PR! Can you also please add a `CHANGELOG.md` entry?

mroswald committed:
following PR feedback

mroswald committed:
update CHANGELOG.md

mroswald committed:
fix lint

mroswald commented:
> Thanks for the bug report and the PR! Can you also please add a `CHANGELOG.md` entry?

Hey @behenate , thanks for taking the time to review this and giving suggestions! I tested your updated version and it works. Added the entry for the change log and fixed one issue the linter found

mroswald committed:
move entry to correct CHANGELOG file

mroswald commented:
Rebased the main branch on my origin force-pushed to trigger a change. Not sure if it affected the tests to run on a wider scope than intended, but I didn't touch the modules that failed. Hope it works now.

behenate commented on packages/expo-print/ios/ExpoPrintWithPrinter.swift:
```diff
@@ -137,11 +138,39 @@ public class ExpoPrintWithPrinter {
-      return try Data(contentsOf: URL(fileURLWithPath: uri))
-    } catch {
-      return nil
+  let printURLSession = URLSession(configuration: .default)
+  var dataTask: URLSessionDataTask?
```
Can you move these to the top of the file next to other variable declarations?

mroswald committed:
move variable def to top of class

tsapeta reviewed (approved):
Looks good to me, thank you for fixing this! We'll release it soon in SDK 49 😉
